File: 54/solution.go
package _54

func spiralOrder(matrix [][]int) []int {
	m, n := len(matrix), len(matrix[0])
	a, b, c, d := 0, n-1, 0, m-1
	res := []int{}
	for a <= b && c <= d {
		for i := a; i <= b; i++ {
			res = append(res, matrix[c][i])
		}
		c++
		for i := c; i <= d; i++ {
			res = append(res, matrix[i][b])
		}
		b--
		if a > b || c > d {
			break
		}
		for i := b; i >= a; i-- {
			res = append(res, matrix[d][i])
		}
		d--
		for i := d; i >= c; i-- {
			res = append(res, matrix[i][a])
		}
		a++
	}
	return res
}

File: 54/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        int a = 0, b = n - 1, c = 0, d = m - 1;
        List<Integer> res = new ArrayList<>();
        while(a <= b && c <= d) {
            for(int i = a; i <= b; i++) {
                res.add(matrix[c][i]);
            }
            c++;
            for(int i = c; i <= d; i++) {
                res.add(matrix[i][b]);
            }
            b--;
            if (a > b || c > d) {
                break;
            }
            for(int i = b; i >= a; i--) {
                res.add(matrix[d][i]);
            }
            d--;
            for(int i = d; i >= c; i--) {
                res.add(matrix[i][a]);
            }
            a++;
        }
        return res;
    }
}